### ISSUE TITLE
getSymbolAtLocation returns undefined even though internal .symbol property exists

### ISSUE THREAD

**Issue by dsherret** [Microsoft/TypeScript#5218]
**Problem (Bug?)**

I've been playing around with the compiler api tonight and I noticed a lot of occasions where `typeChecker.getSymbolAtLocation(node)` returns `undefined`, but accessing the internal `symbol` property of `node` does. Looking briefly at the code for `getSymbolAtLocation`, it doesn't seem to handle many scenarios (or at least not all the ones I would expect, though `getSymbolsInScope` does).

For example:

```typescript
const options: ts.CompilerOptions = { noLib: false, experimentalDecorators: true };
const host = ts.createCompilerHost(options);
const program = ts.createProgram(fileNames, options, host);
const typeChecker = program.getTypeChecker();

console.log(typeChecker.getSymbolAtLocation(program.getSourceFiles()[1])); // undefined
console.log((program.getSourceFiles()[1] as any).symbol); // ok
```

Right now `.symbol` is marked as `internal` on `interface Node` [here](https://github.com/Microsoft/TypeScript/blob/1b5dc0d7d1ecee73cd71019de4a91ce24744c7fb/src/compiler/types.ts#L445).

**Question**

Can `getSymbolAtLocation` be changed to support this scenario? Maybe return `.symbol` when it exists, instead of returning `undefined` in so many scenarios?

I have only been looking at the compiler code tonight for the first time, so maybe I'm completely misunderstanding the purpose of this function and maybe there's something else that's supposed to be used here? The current behaviour seems to be fairly limiting.

**Comment by RyanCavanaugh** [comment]
Can you go into more detail about what you're trying to do? `getSymbolAtLocation` is mainly used for the "Go to Definition" feature in editor scenarios, so it might not be the best function for your use case.

**Comment by dsherret** [comment]
@RyanCavanaugh thanks for replying. I'm trying to get the structure of a file in order to do some automatic code generation based on the class name, method names, parameters, and class & method decorators. I'm automatically generating typed client side code based on server side code in order to access the server (kind of like how service code gets generated automatically in c#)... I think it's a good idea.

Right now I've had some luck creating something very simple by dealing with symbols, but maybe I've gone down the rough path?

**Comment by mhegazy** [comment]
> `console.log(typeChecker.getSymbolAtLocation(program.getSourceFiles()[1])); // undefined` 

why do you want to get the symbol of the sourceFile?

**Comment by dsherret** [comment]
@mhegazy right now, If I have access to `symbol` I can get all the exported members of a file by doing (assuming one class export in a file):

```typescript
const sourceFile: ts.SourceFile = ...;
let exports = sourceFile.symbol.exports;

// or by doing
exports = typeChecker.getExportsOfModule(sourceFile.symbol);

const className = exports[0].getName();
```

*However*, maybe I should be going after declarations instead of symbols? I just found this out... a bit more work though:

```typescript
const sourceFile: ts.SourceFile = ...;
const exportedSymbols = typeChecker.getSymbolsInScope(sourceFile, ts.SymbolFlags.Export);

const className = (exportedSymbols[0].declarations[0].name as ts.LiteralExpression).text;
```

On first look, that seems to have everything I need actually... I think I'm going to close this issue soon because of that, but is this the best way to do this?

I've only been looking at the compiler api recently so sorry for my ignorance :)

**Comment by mhegazy** [comment]
we can definitely add the case of sourceFile to getSymbolAtLocation, though it is not used in our code base; but getSymbolsInScope, or just walking the SourceFile would be better.

**Comment by dsherret** [comment]
I've been looking into this a bit more and it's so much easier to walk the symbols for what I'm doing.

I'm going to commit a sin and in my wrapper class for TypeChecker do this:

```typescript
getSymbolAtLocation(node: ts.Node): ts.Symbol {
    return (node as any).symbol;
}
```

If that eventually breaks then so be it! Just makes it super easy for me in the meantime.

Don't worry about my suggestion. I know you guys have more important things to work on than what I've talked about here. Thanks for the help and clarifications!

I'm going to close this issue, but re-open it if you think other people would want what I suggested with sourcefile.
